Chore: (Docs) Adds test coverage to the docs (#19957)
storybookjs/storybook: docs/writing-tests/test-coverage.md

---
title: 'Test coverage'
---

Test coverage is the practice of measuring whether existing tests fully cover your code. That means surfacing areas which aren't currently being tested, such as: conditions, logic branches, functions and variables.

Coverage tests examine the instrumented code against a set of industry-accepted best practices. They act as the last line of QA to improve the quality of your test suite.

![Test coverage](./test-coverage-storybook.gif)

## Code instrumentation with the coverage addon

Storybook provides an official [test coverage addon](https://storybook.js.org/addons/@storybook/addon-coverage). Powered by [Instanbul](https://istanbul.js.org/), which allows out-of-the-box code instrumentation for the most commonly used frameworks and builders in the JavaScript ecosystem.

### Set up the coverage addon

> To enable test coverage with Storybook, you'll need to take additional steps to set it up properly. We recommend you go through the [test runner documentation](./test-runner.md) before proceeding with the rest of the required configuration.

Reviewer: Maybe make it more explicitly that users should use the test-runner in order to get optimal experience? The addon does help if users want to use something like playwright directly, but it's best suited for test-runner users

| `checkProd`            | Configures the plugin to skip instrumentation in production environments <br/>`options: { instanbul: { checkProd: true,}}`                              | Vite        |
| `forceBuildInstrument` | Configures the plugin to add instrumentation in build mode <br/>`options: { instanbul: { forceBuildInstrument: true,}}`                                 | Vite        |

## What about other coverage reporting tools?

Out of the box, code coverage tests work seamlessly with Storybook's test-runner and the [`@storybook/addon-coverage`](https://storybook.js.org/addons/@storybook/addon-coverage). However, that doesn't mean you can't use additional reporting tools (e.g., [Codecov](https://about.codecov.io/)). For instance, if you're working with [LCOV](https://wiki.documentfoundation.org/Development/Lcov), you can use the generated output (in `coverage/storybook/coverage-storybook.json`) and create your own report with:

<!-- prettier-ignore-start -->

<CodeSnippets
  paths={[
    'common/storybook-coverage-report-lcov.js.mdx',
  ]}
/>

<!-- prettier-ignore-end -->

---

## Troubleshooting

### Run test coverage in other frameworks

> If you intend on running coverage tests in other frameworks (e.g., Vue, Svelte), you'll need to adjust your configuration and enable the required file extensions. For example, if you're using Vue, you'll need to add the following to your nyc configuration file (i.e., `nycrc.json` or `nyc.config.js`):

Reviewer: ```suggestion
If you intend on running coverage tests in frameworks with special files like Vue or Svelte, you'll need to adjust your configuration and enable the required file extensions. For example, if you're using Vue, you'll need to add the following to your nyc configuration file (i.e., `nycrc.json` or `nyc.config.js`):
```
